include/xmp.h

```diff
@@ -320,6 +320,11 @@ struct xmp_frame_info {			/* Current frame information */
 	struct xmp_channel_info channel_info[XMP_MAX_CHANNELS];		/* Current channel information */
 };
 
+struct xmp_callbacks {
+	unsigned long (*read_func)(void *dest, unsigned long len, unsigned long nmemb, void *priv);
+	int           (*seek_func)(void *priv, long offset, int whence);
+	long          (*tell_func)(void *priv);
+};
```

Comment:
Users may want a `close_func` here

Reply:
I think that's a reasonable addition. Do you think it's OK to add an stddef.h include to give `read_func` some `size_t`s instead of `unsigned long`? (It'd only be a "problem" on x86-64 Windows with extremely large files...)

Reply:
If someone crazy truly wants to read more than 4g, then it's a problem - we
don't do that within libxmp. Can you think of any other sources of issue?

As for stddef.h inclusion in xmp.h: I'm not against it, nor am I objecting it - 
only if really necessary. Comments? @cmatsuoka, @sagamusix?

Reply:
I can only speak from libopenmpt perspective where we used stddef.h/cstddef everywhere: short/long/long long etc are all terrible names and as far as I'm concerned they cannot disappear quickly enough.

Reply:
Well, thinking of it, (unsigned) long vs size_t _might_ be problem with modules
in large containers where the tell function would report wrong values.

Reply:
File APIs in general should not depend on pointer size. It is entirely possible to operate on files > 4 GB in a 32-bit process if done correctly. Any offsets really should be (u)int64_ts.

Reply:
That makes us guarantee things.

Then again, the callback api doesn't know anything about the file: the client itself does.
The (unsigned) long use would be safe provided that the module files themselves are
not really 'large' files even if they are contained in large containers.

So, speaking for myself, I'm good with (unsigned) long use, but I don't object to size_t.
(u)int64 on the other hand, would dictate stdint.h.

Reply:
It's 2021. Is someone really going to compile the very latest version of libxmp on a platform that doesn't have stdint.h?

Reply:
> It's 2021. Is someone really going to compile the very latest version of libxmp on a platform that doesn't have stdint.h?

The only toolchain that doesn't provide stdint.h possibly would be MSVC < 2013 (?),
which we can handle by some ugliness in our header.

Reply:
I think at this point, if someone wants to keep using such an old MSVC version, they can be expected to provide their own stdint.h. It was already a common practice some years ago when those were the latest versions of MSVC.

Reply:
@cmatsuoka, @AliceLR, what do you think?

If I have to vote, I vote for keeping (unsigned) long as they are.

Reply:
Another thing—should `xmp_test_module_from_callbacks` ignore `close_func` if it's added? Presumably you would want to provide the same open stream and callbacks to both, but if the callbacks included a `close_func`, you'd either need two sets of callbacks or to reopen the stream...

Reply:
Ugh... Well, if `close_func` is not NULL, I'd think that the user implies that we should close the thing?

Reply:
Okay, now that I've caught up on the backlog (I had this page left open when I commented and GitHub didn't load the conversation from 3 hours ago in...).

My take on `stdint.h` is roughly the same for my own projects but in this case libxmp has done a good job of steering clear of C99-isms, which would cause issues with older MSVCs (which hopefully everyone has updated away from). Adding an include for a C99 header, particularly in a public API header like `xmp.h` seems questionable.

In the case of `seek_func` and `tell_func` I used `long` mainly because the corresponding `stdio.h` functions also use those. Using 64-bit types for offsets would be preferable but if the only good way to do it is `stdint.h` in a public header... ew.

Re: `size_t` I definitely prefer it to `unsigned long` for the case of `read_func`. I think for that particular function that (unlike seek/tell) it's incredibly unlikely that 64-bit values will be required there ever.

Reply:
Even `size_t` won't be necessary if we assume that there are no modules
which are larger than 4GB (see my points at https://github.com/libxmp/libxmp/pull/362#discussion_r646041441)

Reply:
> Even `size_t` won't be necessary if we assume that there are no modules
> which are larger than 4GB (see my points at [#362 (comment)](https://github.com/libxmp/libxmp/pull/362#discussion_r646041441))

Besides, it's no different from `xmp_load_module_from_memory` having a `long size` parameter.
So I vote that we keep the `(unsigned) long`s as they are in this.

Reply:
I'll go ahead and just leave the `long`/`unsigned long` usage then (it can be revisited for 5.0 if it really ends up being a problem, which I doubt...).

Reply:
OK then, the only thing left is the `close_func` being implemented?

Reply:
I think so. (It could probably use some other light cleanup and better documentation too.)

Reply:
I've implemented this function. There was another thing I want to check into, which is that for whatever reason the Prowizard module in the `xmp_test_module_from_callbacks` test wasn't working correctly, and I ended up removing it. Prowizard modules should work fine with this function so I don't know what the deal is.